Increase Application own order among extensions

mihails-strasuns-sociomantic opened the pull request:
To address https://github.com/sociomantic-tsunami/ocean/issues/500

mihails-strasuns-sociomantic commented:
FYI @joseph-wakeling-sociomantic

gavin-norman-sociomantic commented:
Could you add a brief explanation in the commit message of why this change is required?

mihails-strasuns-sociomantic commented:
Sure, probably worth release notes too. For now you can check linked issue though :)

joseph-wakeling-sociomantic commented:
Thanks!  Just tried this out with an app moving much of its resource instantiation to `processConfig`, which fails without this patch, and works with it.

joseph-wakeling-sociomantic commented:
Agree that some "why" is needed in commit message and release notes.

leandro-lucarella-sociomantic reviewed (commented):
Why does it matter? Extensions have a priority which should set the order in which they are executed. Or is this purely for the initialization phase?

leandro-lucarella-sociomantic commented:
(maybe the commit message could clarify that)

mihails-strasuns-sociomantic commented:
> Extensions have a priority which should set the order in which they are executed

How is this supposed to work? Specific case where this matters for #500 is https://github.com/sociomantic-tsunami/ocean/blob/8a7db8ce9fd9eb48d8bca17969ff547651e7d5cf/src/ocean/util/app/ext/ConfigExt.d#L317-L320 - for it to work as expected, application has to be last in that extension array.

joseph-wakeling-sociomantic commented:
@leandro-lucarella-sociomantic consider the case where e.g. you want to instantiate something before event scheduling starts (e.g. something that will be relied on by `onStatsTimer`), and that thing relies on having epoll available.

In principle the natural place to do this is `processConfig`, but in practice right now that's not possible, because `theScheduler.epoll` will not be initialized before `processConfig` is called.

leandro-lucarella-sociomantic commented:
Extension will be sorted according to `order()` when they are `registerExtension()`ed.

https://github.com/sociomantic-tsunami/ocean/blob/8a7db8ce9fd9eb48d8bca17969ff547651e7d5cf/src/ocean/util/app/model/ExtensibleClassMixin.d#L94-L104

leandro-lucarella-sociomantic commented:
So I think the correct fix for this is make your application have a very high `order()` if you want it to execute last. But that will violate some assumptions are made that applications themselves have `order()` 0 (not sure if anything will actually break). The ordering of extensions is probably the trickiest and more annoying part of the app framework. :(

leandro-lucarella-sociomantic commented:
Maybe some scheduler stuff have to be moved to the `preRun()` hook instead of `run()`?

leandro-lucarella-sociomantic commented:
https://github.com/sociomantic-tsunami/ocean/blob/8a7db8ce9fd9eb48d8bca17969ff547651e7d5cf/src/ocean/util/app/model/IApplicationExtension.d#L49-L59

joseph-wakeling-sociomantic commented:
@leandro-lucarella-sociomantic the key thing here is that we need to have some way to instantiate app class resources _after_ config has been parsed and `theScheduler` has been initialized, but _before_ event scheduling actually starts.

Until https://github.com/sociomantic-tsunami/ocean/pull/460 we could just do this kind of stuff in the beginning of `run`, because so long as the stuff got instantiated in less than 30 seconds, there was no issue (which is a flaky solution, but worked in practice).  But now we have actual segfaults happening because `onStatsTimer` tries to use uninstantiated class fields, which can't be instantiated in the app constructor.

mihails-strasuns-sociomantic commented:
> But that will violate some assumptions are made that applications themselves have order() 0

That sounds very wrong to me and what I was trying to fix in this PR. Downstream shouldn't care at all what is internal implementation detail of `DaemonApp`, by the time control reaches user code everything needs to be initialized and ready whenever possible. So application should have something like `int.max` order.

mihails-strasuns-sociomantic commented:
@joseph-wakeling-sociomantic the issue here is not specific to scheduler or config - I want to establish strict guarantee that all extension code runs before app code, one that holds true in all possible situations.

joseph-wakeling-sociomantic commented:
@mihails-strasuns-sociomantic understood and agreed.  Just wanted to make clear the specific problems that we are having to deal with.

leandro-lucarella-sociomantic commented:
> @leandro-lucarella-sociomantic the key thing here is that we need to have some way to instantiate app class resources after config has been parsed and theScheduler has been initialized, but before event scheduling actually starts.

This sounds a lot like we need a new extension point.

leandro-lucarella-sociomantic commented:
> @joseph-wakeling-sociomantic the issue here is not specific to scheduler or config - I want to establish strict guarantee that all extension code runs before app code, one that holds true in all possible situations.

Last time I looked at this code was probably when I wrote it :D, but I think this might not work. I don't remember the details, but I think there are some cases where extension code needs to run after the app, but you can check the `order()` for all extensions to see if that's true. In any case, if it works, the fix would be just to increase the `order()` of the app, not change where the `registerExtension` happens.

Still, with cases like this, when you need something to happen between X and Y no matter which other extensions are used, you probably want to define a new extension point/hook. That can even be done inside the extension iteself, like the config or args extensions do.

joseph-wakeling-sociomantic commented:
I'm starting to feel that @gavin-norman-sociomantic has a definite point that we are overcomplicating our app class framework and that we should simplify down to our actual use cases ... ;-)

Still, I'm happy with any option here as long as we have a way forward.  What I'm worried about is that right now, we can have arbitrary segfaults with our applications if they start up too close to the :00 or :30 points.  Could it be worth rolling that back temporarily until we have our solution in place?

(Memories of Tom: "Never roll back!!"...)

mihails-strasuns-sociomantic commented:
@leandro-lucarella-sociomantic I checked all extensions and the only one right now that uses order above 0 is `VersionArgsExt`. Reading its code I see that the only reason why it may matter is to terminate argument parsing early if `--version` is specified (and not print any errors from app own parsing if present). Not sure if this really any important.

Everything else will work exactly the same with `int.max` order for app.

leandro-lucarella-sociomantic commented:
OK, so do you have any plans going forward then?

mihails-strasuns-sociomantic commented:
Pushed proposed fix.

leandro-lucarella-sociomantic commented on src/ocean/util/app/Application.d:
```diff
@@ -334,13 +334,16 @@ class Application : IApplication
     public override int order ( )
     {
-        return 0;
+        return 1_000_000;
```
I don't like it. I like that the application has order 0, is the "center" of the framework. Why changing this at all?

leandro-lucarella-sociomantic replied:
I see there are a lot of extensions using order 0, I think the correct fix would be to make those extensions use some negative value that makes sense (here is when it gets tricky, to get an order that make the extensions play well between themselves).

joseph-wakeling-sociomantic replied:
@leandro-lucarella-sociomantic we could adopt the BASIC line numbering principle of increments of 10 or more, to leave room for other priorities to be inserted ... :trollface: 

Note that 0 is explicitly described in the `IExtension` doc as the default "if order doesn't matter" setting:
https://github.com/sociomantic-tsunami/ocean/blob/v4.1.1/src/ocean/util/app/model/IExtension.d#L35-L36

So if this is true, it is _not_ the case that the correct `order` for the app extension is 0, because for the app extension the order _does_ matter.

A related question is if it can possibly be true (or reasonable) that an extension could/should have a higher `order` than the application extension.

joseph-wakeling-sociomantic replied:
Note, I'm not anti the idea of 0 being reinterpreted as application-extension priority, so individual extensions should specify priority in line with whether they need to be ready before or after the application extension.  But that means we have to change the meaning of the `order` return value.

joseph-wakeling-sociomantic replied:
Reading the docs of `Application.d` it looks like we have a clash here between what is specified in `IExtension.order` doc and what is specified in `Application` doc.

mihails-strasuns-sociomantic replied:
> I don't like it. I like that the application has order 0, is the "center" of the framework. Why changing this at all?

Because `0` is the default value for extension order and we need application to have lower order by default. And relying on virtue of developers to always modify order value of new added extensions is not going to work.

We can keep app at 0 if base `order` method return value for all extensions is defined to `-1_000` or something like that.

leandro-lucarella-sociomantic replied:
There is no default, you have to always explicitly set one. To be honest, I can understand that maybe the docs are difficult to navigate, but it's all there. I think this is a docs-only issue, so maybe we can do that, recommend some value for extensions (like `-1_000`), saying that extensions want to almost always run before the application.

mihails-strasuns-sociomantic replied:
That sounds exactly the same as setting app to `1_000` but (considerably) more error-prone. Why is this so important? Application framework is convoluted enough that at this point you can safely assume no on reads docs when writing extensions, thus suitable defaults are extremely important.

leandro-lucarella-sociomantic replied:
It's not important but minimizes the chances that we fuck up something else. For example your PR didn't update the Application docs. There might be other areas where the App having order of 0 is assumed. I don't see a strong point to change it, and without a strong point, I prefer to minimize the chances of fuck ups.

leandro-lucarella-sociomantic replied:
Specially for a patch-release for me this makes NO sense. If it were a major (or maybe a minor) then maybe. I don't know if other apps might be relying on this order.

mihails-strasuns-sociomantic replied:
Right now it is semi-undefined (in a sense that most extensions use order 0 and resulting order ends up being "reverse order of register calls") - hard to rely on this.

Will you be OK with having this in a minor?

leandro-lucarella-sociomantic replied:
> Right now it is semi-undefined (in a sense that most extensions use order 0 and resulting order ends up being "reverse order of register calls") - hard to rely on this.

Well, for those extensions order was important, so I would say is a bug in those extensions and those extensions should be fixed (change the order to some negative value). For me that's the fix that should go in a patch release.

> Will you be OK with having this in a minor?

Change the Application order to something bigger than 0? I'll be OK-ish, but I strongly would recommend against it, because is a silent change on behaviour for any project deriving from `Application` that could have unpredictable and hard to track consequences. I think that sort of changes belong to a major.

leandro-lucarella-sociomantic replied:
@gavin-norman-sociomantic it would be useful to have your POV on this.

gavin-norman-sociomantic replied:
Given that:
1. I think we should seriously look into a major refactoring of the internals of the app framework soon,
2. it's pretty unlikely that more extensions will be added any time soon,

I think it makes the most sense to keep the meaning of 0 as the natural middle point (app initialisation), and move all extensions that should happen before then to a negative order.

leandro-lucarella-sociomantic commented:
https://github.com/sociomantic-tsunami/ocean/blob/b200e555190591538c44ca789c5dc26fc2540928/src/ocean/util/app/Application.d#L39-L43

joseph-wakeling-sociomantic commented:
If this is going to be contentious, perhaps we should bump it to a later release and allow v3.7.2 to go out ... ?  There is useful stuff in there.

mihails-strasuns-sociomantic committed:
Make most extensions have lower order than app

Previously several extensions used in DaemonApp had same order as application
itself meaning it was not possible to rely on `processConfig` method being
run after `processConfig` of those extensions. That proved to be an extra
burden, particularly with `TaskExt` and now all extensions are ordered before
the application itself unless it is strictly necessary otherwise.

mihails-strasuns-sociomantic commented:
Updated

gavin-norman-sociomantic commented on relnotes/order.feature.md:
```diff
@@ -0,0 +1,7 @@
+### Changed order of app extensions
+
+Previously several extensions used in DaemonApp had same order as application
```
Maybe put this description in the commit message too?

mihails-strasuns-sociomantic replied:
Sure, done
